teuthology/nuke.py

```diff
@@ -361,6 +367,101 @@ def synch_clocks(remotes):
+                     " because it was created " + str(i.get_created()) +
+                     " seconds ago which is older than " +
+                     str(OPENSTACK_DELAY) + " and it is not locked")
+            if not ctx.dry_run:
+                i.destroy()
+            continue
+        log.debug("stale-openstack: instance " + i['name'] + " OK")
+
+def openstack_delete_volume(id):
+    sh("openstack volume delete " + id + " || true")
+
+def stale_openstack_volumes(ctx, volumes):
+    now = datetime.datetime.now()
+    for volume in volumes:
+        volume = json.loads(sh("openstack volume show -f json " +
+                               volume['ID']))
+        volume = dict(map(lambda v: (v['Field'], v['Value']), volume))
+        created_at = datetime.datetime.strptime(
+            volume['created_at'], '%Y-%m-%dT%H:%M:%S.%f')
+        created = (now - created_at).total_seconds()
+        if created > ctx.teuthology_config['max_job_time'] + OPENSTACK_DELAY:
+            log.info("stale-openstack: volume " + volume['display_name'] +
+                     "(" + volume['id'] + ") destroyed because it"
+                     " was created " + str(created) +
+                     " seconds ago which is older than max_job_time " +
+                     str(ctx.teuthology_config['max_job_time']) + " + " +
+                     str(OPENSTACK_DELAY))
+            if not ctx.dry_run:
+                openstack_delete_volume(volume['id'])
+            continue
+        log.debug("stale-openstack: volume " + volume['id'] + " OK")
+
+def stale_openstack_nodes(ctx, instances, locked_nodes):
+    for (name, node) in locked_nodes.iteritems():
+        name = decanonicalize_hostname(name)
+        if node['machine_type'] != 'openstack':
+            continue
+        if (name not in instances and
+            locked_since_seconds(node) > OPENSTACK_DELAY):
+            log.info("stale-openstack: node " + name + " unlocked" +
+                     " because it was created " +
+                     str(locked_since_seconds(node)) +
+                     " seconds ago which is older than " +
+                     str(OPENSTACK_DELAY) + " and it has no instance")
+            if not ctx.dry_run:
+                unlock_one(ctx, name, node['locked_by'])
+            continue
+        log.debug("stale-openstack: node " + name + " OK")
+
+def openstack_remove_again():
+    """Volumes and servers with REMOVE-ME in the name are leftover
+    that failed to be removed. It is not uncommon for a failed removal
+    to succeed later on."""
+    sh("""
+    openstack server list --name REMOVE-ME --column ID --format value |
+    xargs --no-run-if-empty --max-args 1 -P20 openstack server delete --wait
+    true
+    """)
+    sh("""
+    openstack volume list --name REMOVE-ME --column ID --format value |
+    xargs --no-run-if-empty --max-args 1 -P20 openstack volume delete
+    true
+    """)
+
+def stale_openstack(ctx):
```

Comment:
Could this be moved above the other functions it calls? It'll help with future readability